itou/job_applications/enums.py

```diff
@@ -15,3 +15,52 @@ def sender_kind_to_pe_origine_candidature(sender_kind):
         SenderKind.PRESCRIBER: "PRES",
         SenderKind.SIAE_STAFF: "EMPL",
     }.get(sender_kind, "DEMA")
+
+
+class RefusalReason(models.TextChoices):
+    DID_NOT_COME = "did_not_come", "Candidat non joignable"
+    DID_NOT_COME_TO_INTERVIEW = "did_not_come_to_interview", "Candidat ne s’étant pas présenté à l’entretien"
+    HIRED_ELSEWHERE = "hired_elsewhere", "Candidat indisponible : en emploi"
+    TRAINING = "training", "Candidat indisponible : en formation"
+    NOT_INTERESTED = "not_interested", "Candidat non intéressé"
+    NON_ELIGIBLE = "non_eligible", "Candidat non éligible"
+    NOT_MOBILE = "not_mobile", "Candidat non mobile"
+    LACKING_SKILLS = "lacking_skills", "Le candidat n’a pas les compétences requises pour le poste"
+    INCOMPATIBLE = "incompatible", "Un des freins à l'emploi du candidat est incompatible avec le poste proposé"
+    PREVENT_OBJECTIVES = (
+        "prevent_objectives",
+        "L'embauche du candidat empêche la réalisation des objectifs du dialogue de gestion",
+    )
+    NO_POSITION = "no_position", "Pas de recrutement en cours"
+    OTHER = "other", "Autre (détails dans le message ci-dessous)"
+
+    # Hidden reasons kept for history.
+    APPROVAL_EXPIRATION_TOO_CLOSE = (
+        "approval_expiration_too_close",
+        "La date de fin du PASS IAE / agrément est trop proche",
+    )
+    UNAVAILABLE = "unavailable", "Candidat indisponible ou non intéressé par le poste"
+    ELIGIBILITY_DOUBT = (
+        "eligibility_doubt",
+        "Doute sur l'éligibilité du candidat (penser à renvoyer la personne vers un prescripteur)",
+    )
+    DEACTIVATION = "deactivation", "La structure n'est plus conventionnée"
+    POORLY_INFORMED = "poorly_informed", "Candidature pas assez renseignée"
+
+    @classmethod
+    def hidden(cls):
+        """Old refusal reasons kept for history but not displayed to end users."""
+        return [
+            cls.APPROVAL_EXPIRATION_TOO_CLOSE,
+            cls.DEACTIVATION,
+            cls.ELIGIBILITY_DOUBT,
+            cls.UNAVAILABLE,
+            cls.POORLY_INFORMED,
+        ]
+
+    # I wasn't able to override parent's `choice`` method so here's a workaround.
```

Comment:
Ça m'étonnais mais la surcharge d'une `property` fonctionne comme on pourrais si attendre :
```python
In [62]: class A:
    ...:     @property
    ...:     def prop(self):
    ...:         return 'A'
    ...: 

In [63]: class B(A):
    ...:     @property
    ...:     def prop(self):
    ...:         return 'B'
    ...: 

In [64]: a = A()

In [65]: b = B()

In [66]: a.prop
Out[66]: 'A'

In [67]: b.prop
Out[67]: 'B'
```
Mais parce que `models.TextChoices` est un `enum` [1] et que Django crée la property dans une metaclasse [2], en surchargeant ici tu ne fait que ajouter la `property` aux membres de l'enum sauf que `.choices` est appelé sur la classe et pas un membre de l'enum :
```python
In [74]: class C(models.TextChoices):
    ...:     A = "A", "Text A"
    ...:     B = "B", "Text B"
    ...:     @property
    ...:     def choices(self):
    ...:         return ["Nope"]
    ...: 

In [75]: C.choices
Out[75]: [('A', 'Text A'), ('B', 'Text B')]

In [76]: C.A.choices
Out[76]: ['Nope']
```
Je n'ai pas essayer mais ça devrait donc pouvoir fonctionner en surchargeant au même niveau que Django donc dans une metaclass.


[1] https://docs.python.org/3/library/enum.html#allowed-members-and-attributes-of-enumerations
[2] https://github.com/django/django/blob/main/django/db/models/enums.py#L59

Reply:
Romain, t'es un génie ! <3
J'ai réussi à faire ce que je voulais grâce à toi ! J'avoue que je n'ai pas encore tout à fait compris pourquoi ça fonctionne, mais ça fonctionne. J'aimerais creuser le fonctionnement des Metaclasses la semaine prochaine. Encore merci !!!